internal/controller/pipelinerollout_controller.go

```diff
@@ -731,28 +731,62 @@ func (r *PipelineRolloutReconciler) updatePipelineRolloutStatusToFailed(ctx cont
 	return r.updatePipelineRolloutStatus(ctx, pipelineRollout)
 }
 
-func (r *PipelineRolloutReconciler) makePipelineDefinition(ctx context.Context, pipelineRollout *apiv1.PipelineRollout) (*kubernetes.GenericObject, error) {
-	labels, err := pipelineLabels(pipelineRollout)
+// getPipelineName retrieves the name of the current running pipeline managed by the given
+// pipelineRollout through the `promoted` label. Unless there is non such pipeline exist, then
+// construct the name by calculate the suffix and append to the PipelineRollout name.
+func (r *PipelineRolloutReconciler) getPipelineName(ctx context.Context, pipelineRollout *apiv1.PipelineRollout) (string, error) {
+	pipelines, err := kubernetes.ListCR(
+		ctx, r.restConfig, common.NumaflowAPIGroup, common.NumaflowAPIVersion, "pipelines",
+		pipelineRollout.Namespace, fmt.Sprintf(
+			"%s=%s,%s=%s", common.LabelKeyPipelineRolloutForPipeline, pipelineRollout.Name,
+			common.LabelKeyPipelineRolloutForPipeline, common.LabelValueUpgradePromoted,
+		), "")
 	if err != nil {
-		return nil, err
+		return "", err
+	}
+	if len(pipelines) > 1 {
+		return "", fmt.Errorf("there should only be one promoted pipeline")
+	} else if len(pipelines) == 0 {
+		suffixName, err := r.calPipelineNameSuffix(ctx, pipelineRollout)
+		if err != nil {
+			return "", err
+		}
+		return pipelineRollout.Name + suffixName, nil
 	}
+	return pipelines[0].Name, nil
+}
 
+// calPipelineNameSuffix calculates the suffix of the pipeline name by utilizing the `NameCount`
+// field.
+func (r *PipelineRolloutReconciler) calPipelineNameSuffix(ctx context.Context, pipelineRollout *apiv1.PipelineRollout) (string, error) {
 	if pipelineRollout.Status.NameCount == nil {
 		pipelineRollout.Status.NameCount = new(int32)
 		statusUpdateErr := r.updatePipelineRolloutStatus(ctx, pipelineRollout)
 		if statusUpdateErr != nil {
-			return nil, statusUpdateErr
+			return "", statusUpdateErr
 		}
 	}
 
-	curNameCount := *pipelineRollout.Status.NameCount
+	return "-" + fmt.Sprint(*pipelineRollout.Status.NameCount), nil
+}
+
+func (r *PipelineRolloutReconciler) getPipelineDefinition(ctx context.Context, pipelineRollout *apiv1.PipelineRollout) (*kubernetes.GenericObject, error) {
```

Comment:
nit: do we rename it? The new name makes it sound like we're finding the definition of the existing pipeline when we're actually creating a new definition - do you agree? I'm okay with adding a comment to clarify if you prefer this name.

Reply:
yeah, let me keep the existing naming